### PR TITLE
PIN-4370 BKE - Tenant process added route to add certified attribute

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala [hunk 1]
```diff
@@ -387,6 +387,58 @@ final case class TenantApiServiceImpl(
...
+        .toFuture(TenantIsNotACertifier(requesterTenantUuid))
+      attribute           <- attributeRegistryManagementService
+        .getAttributeById(seed.id)
+      _                   <- attribute.kind match {
+        case Certified => Future.unit
+        case _         => Future.failed(RegistryAttributeIdNotFound(attribute.id))
+      }
+      targetTenant        <- tenantManagementService.getTenantById(targetTenantUuid).map(_.toManagement)
+      attribute = targetTenant.attributes.flatMap(_.certified).find(_.id == seed.id)
+      updatedTenant <- attribute.fold(
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: Missing a check: an attribute should be added to another tenant only if the attribute's origin field is equal to the requester's "certifierId"

**Review by beetlecrunch — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala [hunk 2]
```diff
@@ -387,6 +387,70 @@ final case class TenantApiServiceImpl(
...
+              attribute.origin,
+              certifierId
+            )
+          )
+      }
+      targetTenant        <- tenantManagementService.getTenantById(targetTenantUuid).map(_.toManagement)
+      attribute = targetTenant.attributes.flatMap(_.certified).find(_.id == seed.id)
+      updatedTenant <- attribute.fold(
+        tenantManagementService.addTenantAttribute(targetTenantUuid, seed.toCreateDependency(now))
+      )(_ => Future.failed(CertifiedAttributeAlreadyExists(targetTenantUuid, seed.id)))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: here we may be in the case where the caller wants to re-assign a revoked attribute.
I think it should be allowed

File: src/main/scala/it/pagopa/interop/tenantprocess/error/ResponseHandlers.scala [hunk 3]
```diff
@@ -108,6 +108,18 @@ object ResponseHandlers extends AkkaResponses {
...
+  def addCertifiedAttributeResponse[T](logMessage: String)(
+    success: T => Route
+  )(result: Try[T])(implicit contexts: Seq[(String, String)], logger: LoggerTakingImplicit[ContextFieldsToLog]): Route =
+    result match {
+      case Success(s)                                                     => success(s)
+      case Failure(ex: CertifiedAttributeOriginIsNotComplaintToCertifier) => forbidden(ex, logMessage)
+      case Failure(ex: TenantIsNotACertifier)                             => forbidden(ex, logMessage)
+      case Failure(ex: TenantByIdNotFound)                                => notFound(ex, logMessage)
+      case Failure(ex: CertifiedAttributeAlreadyExists)                   => conflict(ex, logMessage)
+      case Failure(ex)                                                    => internalServerError(ex, logMessage)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/error/ResponseHandlers.scala]: also
```
case Failure(ex: RegistryAttributeIdNotFound) => badRequest(ex, logMessage)
```

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala [hunk 4]
```diff
@@ -387,6 +387,70 @@ final case class TenantApiServiceImpl(
     }
   }
 
+  override def addCertifiedAttribute(tenantId: String, seed: CertifiedTenantAttributeSeed)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    toEntityMarshallerTenant: ToEntityMarshaller[Tenant]
+  ): Route = authorize(ADMIN_ROLE) {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]:
```suggestion
  ): Route = authorize(ADMIN_ROLE, M2M_ROLE) {
```
@beetlecrunch right?

File: src/main/scala/it/pagopa/interop/tenantprocess/error/TenantProcessErrors.scala [hunk 5]
```diff
@@ -94,4 +94,16 @@ object TenantProcessErrors {
         "0022",
         s"Certified Attribute ($attributeOrigin, $attributeCode) already in tenant $tenantId"
       )
+  final case class CertifiedAttributeAlreadyExists(tenantId: UUID, attributeId: UUID)
+      extends ComponentError("0023", s"Certified Attribute $attributeId already exists in tenant $tenantId")
+
+  final case class CertifiedAttributeOriginIsNotComplaintToCertifier(
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/error/TenantProcessErrors.scala]:
```suggestion
  final case class CertifiedAttributeOriginIsNotCompliantWithCertifier(
```

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala (hunk 4 above)
Comment (beetlecrunch) [reply src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: Yes, I wasn't thinking about the m2m flow.

**Review by beetlecrunch — APPROVED** [review]

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala [hunk 6]
```diff
@@ -423,10 +423,12 @@ final case class TenantApiServiceImpl(
...
       }
       targetTenant        <- tenantManagementService.getTenantById(targetTenantUuid).map(_.toManagement)
-      attribute = targetTenant.attributes.flatMap(_.certified).find(_.id == seed.id)
-      updatedTenant <- attribute.fold(
+      certifiedAttribute = targetTenant.attributes.flatMap(_.certified).find(_.id == seed.id)
+      updatedTenant <- certifiedAttribute.fold(
         tenantManagementService.addTenantAttribute(targetTenantUuid, seed.toCreateDependency(now))
-      )(_ => Future.failed(CertifiedAttributeAlreadyExists(targetTenantUuid, seed.id)))
+      )(_ =>
+        tenantManagementService.updateTenantAttribute(targetTenantUuid, attribute.id, attribute.toCertifiedSeed(now))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: @beetlecrunch should we use the same approach as the revocation and fail if the attribute is already assigned but not revoked?

**Review by galales — COMMENTED** [review]
LGTM, minor detail

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala (hunk 6 above)
Comment (nttdata-rtorsoli) [reply src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: @galales can you check this point again?

File: src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala [hunk 7]
```diff
@@ -426,8 +426,10 @@ final case class TenantApiServiceImpl(
       certifiedAttribute = targetTenant.attributes.flatMap(_.certified).find(_.id == seed.id)
       updatedTenant <- certifiedAttribute.fold(
         tenantManagementService.addTenantAttribute(targetTenantUuid, seed.toCreateDependency(now))
-      )(_ =>
-        tenantManagementService.updateTenantAttribute(targetTenantUuid, attribute.id, attribute.toCertifiedSeed(now))
+      )(attr => attr.revocationTimestamp match {
+        case None => Future.failed(CertifiedAttributeNotRevoked(targetTenantUuid, attribute.id))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/api/impl/TenantApiServiceImpl.scala]: on the business side the name you used initially was more clear
```suggestion
        case None => Future.failed(CertifiedAttributeAlreadyAssigned(targetTenantUuid, attribute.id))
```

File: src/main/scala/it/pagopa/interop/tenantprocess/error/TenantProcessErrors.scala [hunk 8]
```diff
@@ -103,4 +103,10 @@ object TenantProcessErrors {
         "0023",
         s"Organization ${requesterId.toString} not allowed to assign certified attributes to tenant ${tenantId.toString} -> origin $origin , certifier $certifierId"
       )
+
+  final case class CertifiedAttributeNotRevoked(tenantId: UUID, attributeId: UUID)
+      extends ComponentError(
+        "0024",
+        s"Certified Attribute $attributeId not found in tenant $tenantId"
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/tenantprocess/error/TenantProcessErrors.scala]:
```suggestion
        s"Certified Attribute $attributeId already assigned to tenant $tenantId"
```

**Review by galales — APPROVED** [review]
LGTM, just minor rename required

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**